### PR TITLE
[BACKLOG-17102] add request id

### PR DESCRIPTION
**Pull request by chiltonj**
@pentaho/rebelalliance @pentaho/rebelalliance  In conjunction with:  https://github.com/pentaho/pentaho-platform-ee/pull/1096

### CODE REVIEW

File: extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.UUID;
+
+public class RequestIdFilter implements Filter {
+
+  private static final Log logger = LogFactory.getLog( RequestIdFilter.class );
+  public static final String X_REQUEST_ID = "x-request-id";
```
Comment (pentaho-mzur) [inline extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java]: Should these two constants be better placed in ActionUtil class?

File: extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      logger.info( "received request with request id of: " + requestId );
+      MDC.put( REQUEST_ID, "rid-" + requestId );
+
+      chain.doFilter( req, resp );
+
+    } finally {
+
+      logger.info( "Exiting request with request id of: " + requestId );
+      ( (HttpServletResponse) resp ).setHeader( X_REQUEST_ID, requestId );
+      MDC.clear();
```
Comment (pentaho-mzur) [inline extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java]: Why do we clear all entries of the MDC ,while our filter is just concerned about REQUEST_ID key/entry ?

Comment (chiltonj) [inline extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java]: MDC is stored in TLS -- may as well clear on exiting filter chain. good hygiene.

Comment (pentaho-mzur) [inline extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java]: Why?  you don't know your position in the filter chain, right? - there might be parent filter that wants the MDC preserved.

File: extensions/src/main/java/org/pentaho/platform/web/http/api/resources/ActionResource.java [hunk 3]
```diff
@@ -145,6 +149,8 @@ IAction createActionBean( final String actionClass, final String actionId ) thro
 
     public void run() {
       try {
+        Optional.ofNullable( mdcContextMap ).ifPresent( s -> MDC.setContextMap( mdcContextMap ) );
```
Comment (pentaho-mzur) [inline extensions/src/main/java/org/pentaho/platform/web/http/api/resources/ActionResource.java]: somewhat more self-explanatory lambda, that makes it less cryptic?... for your consideration:

        Optional.ofNullable( mdcContextMap ).ifPresent( map -> MDC.setContextMap( map ) );

File: extensions/src/main/java/org/pentaho/platform/web/http/api/resources/ActionResource.java [hunk 4]
```diff
@@ -145,6 +149,8 @@ IAction createActionBean( final String actionClass, final String actionId ) thro
...
+        logger.info( "Running action: " + actionId );
```
Comment (pentaho-amessier) [inline extensions/src/main/java/org/pentaho/platform/web/http/api/resources/ActionResource.java]: Log messages need to be externalized

File: extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  public void doFilter( ServletRequest req, ServletResponse resp, FilterChain chain ) throws ServletException, IOException {
+
+    HttpServletRequest request = (HttpServletRequest) req;
+    String requestId = Optional.ofNullable( request.getHeader( X_REQUEST_ID ) ).orElse( UUID.randomUUID().toString() );
+
+    try {
+
+      logger.info( "received request with request id of: " + requestId );
```
Comment (pentaho-amessier) [inline extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java]: Log messages need to be externalized

File: extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      MDC.put( REQUEST_ID, "rid-" + requestId );
+
+      chain.doFilter( req, resp );
+
+    } finally {
+
+      logger.info( "Exiting request with request id of: " + requestId );
```
Comment (pentaho-amessier) [inline extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java]: Log messages need to be externalized

File: extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java (hunk 1 above)
Comment (pentaho-amessier) [inline extensions/src/main/java/org/pentaho/platform/web/http/filters/RequestIdFilter.java]: I agree

**Pull request merged**